File: src/de/hsh/Julian/Leertastenklatsche.java
package de.hsh.Julian;

import common.config.WindowConfig;
import common.util.Logger;
import common.util.PlaySound;
import de.hsh.dennis.model.NpcLogic.SpawnTimer;
import javafx.application.Platform;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;

import java.util.ArrayList;
import java.util.Observable;
import java.util.Observer;

public class Leertastenklatsche extends Observable implements Observer {


    static final         String     location      = "/de/hsh/Julian/";
    private static final String     soundLocation = "/de/hsh/Julian/wav/";
    private static final SpawnTimer timer         = new SpawnTimer();

    private int              score     = 0;
    private int              leben     = 3;
    private boolean          gamedone  = false;
    boolean horrorWasActivated = false;
    private TheDude          thedude;
    private ArrayList<Enemy> enemyList = new ArrayList<>();
    private double spawntimer;

    public Leertastenklatsche() {
        initGame();
    }

    private void initGame() {
        thedude = new TheDude( WindowConfig.window_width / 2 - 75,
                               WindowConfig.window_height * 0.73
        );
        thedude.setSpeed( 0 );



    }

    void render( Canvas gc ) {
        if ( gamedone ) {
            return;
        }
        createNewEnemies();
        updateEnemies();
        Platform.runLater( () -> {
            renderScore( gc );
            enemyList.forEach( enemy -> {
                enemy.draw( gc );
            } );
            thedude.draw( gc );
        } );
    }

    private void createNewEnemies() {
        if(score >= 70)
            spawntimer= 0.2;
        else
            spawntimer = 2.0d - score / 50;

        if ( timer.elabsedTime() > spawntimer) {
            timer.resetTimer();
            final Enemy e = Enemy.createEnemy();
            e.addObserver( this );
            thedude.addCollidingActor( e );
            enemyList.add( e );
        }
    }

    private void updateEnemies() {
        for ( Enemy enemy : enemyList ) {
            if ( enemy.getX() > WindowConfig.window_width / 2 ) {
                enemy.setPos( enemy.getX() - 1.0 - score / Math.sqrt( score + 1 ), enemy.getY() );
            }
            else {
                enemy.setPos( enemy.getX() + 1 + score / Math.sqrt( score + 1 ), enemy.getY() );
            }
        }
    }

    private void renderScore( Canvas canvas ) {
        GraphicsContext gc         = canvas.getGraphicsContext2D();
        String          pointsText;
        if(score<70) {
            pointsText    ="LEERTASTENKLATSCHE\nGegner abgewehrt: " + (score) + "\nVerbleibende Leben: " + (leben);
            //Logger.log(getClass()+" Score: "+score);

            gc.fillText(pointsText, 360, 36);
            gc.strokeText(pointsText, 360, 36);
        }
        else{
            if(!horrorWasActivated){
                //PlaySound.playSound( "src\\de\\hsh\\Julian\\wav\\horror.wav" );
                PlaySound.playSound( soundLocation + "Kalinka.mp3" );
                leben = 70;
            }

            horrorWasActivated=true;
            pointsText = "HORRORMODUS SCORE: " + (score) + "\nKLATSCH KLATSCH KLATSCH Leben: " + (leben);

            //Logger.log(getClass()+" Score: "+score);

            gc.fillText( pointsText, 360, 36);
            gc.strokeText( pointsText, 360, 36 );

        }
    }

    void parseInput( final String code ) {
        switch ( code ) {
            case "LEFT":
                if ( !thedude.turnedleft ) {
                    thedude.swapImage();
                    thedude.turnedleft = true;
                }
                break;
            case "RIGHT":
                if ( thedude.turnedleft ) {
                    thedude.swapImage();
                    thedude.turnedleft = false;
                }
                break;
            //EASTEREGG ;-))
            case "SPACE":
                if ( !horrorWasActivated ) { PlaySound.playSound( soundLocation + "cat.wav" ); }
                break;
        }
    }

    @Override
    public void update( Observable o, Object arg ) {
        if ( gamedone ) {
            Logger.log( this.getClass() + ": Game should be done, but event is still fired. : " + o + ", " + arg );
            return;
        }
        if ( o instanceof Enemy ) {
            Enemy e = (Enemy) o;
            for ( Enemy enemy : this.enemyList ) {
                if ( enemy.id == e.id ) {
                    //Logger.log( this.getClass() + ": Searched id : " + e.id + " Enemy id : " + enemy.id );

                    e.deleteObservers();
                    enemy.deleteObservers();

                    this.thedude.removeCollisionActor( enemy );
                    enemyList.remove( enemy );

                    //Logger.log( this.getClass() + ": Found enemy with same id" );
                    if ( enemy.getPos()[ 0 ] <= thedude.getPos()[ 0 ] ) {
                        if ( thedude.turnedleft ) {
                            score++;
                            if ( !horrorWasActivated ) { PlaySound.playAndResetSound( soundLocation + "collision.wav" ); }
                        }
                        else {
                            leben--;
                            if ( !horrorWasActivated ) { PlaySound.playAndResetSound( soundLocation + "hit.wav" ); }
                            gameOver();
                        }
                    }
                    else if ( enemy.getPos()[ 0 ] > thedude.getPos()[ 0 ] ) {
                        if ( !thedude.turnedleft ) {
                            score++;
                            if ( !horrorWasActivated ) { PlaySound.playAndResetSound( soundLocation + "collision.wav" ); }
                        }
                        else {
                            leben--;
                            if ( !horrorWasActivated ) { PlaySound.playAndResetSound( soundLocation + "hit.wav" ); }
                            gameOver();
                        }
                    }
                    enemy = null;
                    e = null;

                    return;
                }
            }
        }


    }

    private void gameOver() {

        if ( leben <= 0 && !gamedone ) {
            gamedone = true;
            PlaySound.playSound( soundLocation + "noo.wav" );
            for ( Enemy e : this.enemyList ) {
                e.deleteObservers();
            }
            this.enemyList.removeAll( this.enemyList );
            setChanged();
            notifyObservers( "gameover" );
        }
    }

    public int getScore() {
        return score;
    }

    public void setScore( int score ) {
        this.score = score;
    }
}

File: Games/Julian/src/de/hsh/Julian/Leertastenklatsche.kt
package de.hsh.Julian

import common.config.WindowConfig
import common.util.Logger
import common.util.PlaySound.playAndResetSound
import common.util.PlaySound.playSound
import de.hsh.dennis.model.Spawn.SpawnTimer
import javafx.application.Platform
import javafx.scene.canvas.Canvas
import javafx.scene.input.KeyCode
import java.util.*
import kotlin.math.sqrt

class Leertastenklatsche(observer: Observer) : Observable(), Observer {

    var horrorWasActivated = false
    private var score = 0
    private var leben = 3
    private var gamedone = false
    private var thedude: TheDude = TheDude(
            WindowConfig.window_width / 2.0 - 75.0,
            WindowConfig.window_height * 0.73)
    private val enemyList = ArrayList<Enemy>()
    private var spawntimer = 0.0

    init { this.addObserver(observer) }

    fun render(gc: Canvas) {
        if (gamedone) return

        spawntimer = if (score >= 70) 0.2 else 2.0 - score / 50

        if (timer.elabsedTime() > spawntimer) {
            timer.resetTimer()
            val e = Enemy(this)
            thedude.addCollidingActor(e)
            enemyList.add(e)
        }
        enemyList.forEach { enemy ->
            val enemyX = when(enemy.x > WindowConfig.window_width / 2) {
                true -> enemy.x - 1f - score / sqrt(score + 1f)
                false -> enemy.x + 1f + score / sqrt(score + 1f)
            }
            enemy.setPos(enemyX, enemy.y)
        }
        Platform.runLater {
            renderScore(gc)
            enemyList.forEach { enemy: Enemy -> enemy.draw(gc) }
            thedude.draw(gc)
        }
    }

    private fun renderScore(canvas: Canvas) {
        if (score >= 70) {
            if (!horrorWasActivated) {
                //PlaySound.playSound( "src\\de\\de.hsh\\Julian\\wav\\horror.wav" );
                playSound(pathToSound + "Kalinka.mp3")
                leben = 70
            }
            horrorWasActivated = true
        }
        with(canvas.graphicsContext2D) {
            val pointsText: String =
                    if (score < 70) "LEERTASTENKLATSCHE\nGegner abgewehrt: $score\nVerbleibende Leben: $leben"
                    else "HORRORMODUS SCORE: $score\nKLATSCH KLATSCH KLATSCH Leben: $leben"
            fillText(pointsText, 360.0, 36.0)
            strokeText(pointsText, 360.0, 36.0)
        }
    }

    fun parseInput(code: KeyCode) {
        when (code) {
            KeyCode.LEFT -> if (!thedude.isTurnedLeft) {
                thedude.swapImage()
                thedude.isTurnedLeft = true
            }
            KeyCode.RIGHT -> if (thedude.isTurnedLeft) {
                thedude.swapImage()
                thedude.isTurnedLeft = false
            }
            KeyCode.SPACE -> if (!horrorWasActivated) playSound(pathToSound + "cat.wav")
        }
    }

    override fun update(o: Observable, arg: Any) {
        fun adjustPositions(enemyXPosition: Double) {
            if (enemyXPosition <= thedude.pos[0])
                if (thedude.isTurnedLeft) {
                    score++
                    if (!horrorWasActivated) playAndResetSound(pathToSound + "collision.wav")
                } else {
                    leben--
                    if (!horrorWasActivated) playAndResetSound(pathToSound + "hit.wav")
                    gameOver()
                }
            else if (!thedude.isTurnedLeft)
                score++
                if (!horrorWasActivated) playAndResetSound(pathToSound + "collision.wav")
            else
                leben--
                if (!horrorWasActivated) playAndResetSound(pathToSound + "hit.wav")
                gameOver()
        }

        if (gamedone) {
            Logger.log(this.javaClass.toString() + ": Game should be done, but event is still fired. : " + o + ", " + arg)
            return
        }
        if (o !is Enemy) return

        enemyList.filter { enemy -> enemy.id == o.id }
                .forEach { enemy ->
                    o.deleteObservers()
                    enemy.deleteObservers()
                    thedude.removeCollisionActor(enemy)
                    enemyList.remove(enemy)
                    adjustPositions(enemy.pos[0])
                }
    }

    private fun gameOver() {
        if (leben <= 0 && !gamedone) {
            gamedone = true
            playSound(pathToSound + "noo.wav")
            enemyList.forEach { e ->
                e.deleteObservers()
            }
            enemyList.removeAll(enemyList)
            setChanged()
            notifyObservers("gameover")
        }
    }

    companion object {
        const val location = "/de/hsh/Julian/"
        const val pathToSound = location + "wav/"
        private val timer = SpawnTimer()
    }
}
